File: hash/swift/RosettaImageHash/Sources/RosettaImageHash/Internal/ImgRGB.swift
import Foundation

/// Normalize an RGBImage (3 or 4 channels) to a flat [UInt8] of RGB triples
/// (row-major, length = width * height * 3).
///
/// For .rgba, alpha is composited against opaque black via
/// `out_c = floor(src_c * alpha / 255)`, matching PIL `convert('RGB')`.
/// Fully transparent → (0,0,0); fully opaque → src RGB unchanged.
enum ImgRGB {
	struct RGB3 {
		let width: Int
		let height: Int
		let data: [UInt8]
	}

	static func toRGB(_ image: RGBImage) -> RGB3 {
		if image.channels == .rgb {
			return RGB3(width: image.width, height: image.height, data: image.data)
		}
		// .rgba
		let n = image.width * image.height
		var out = [UInt8](repeating: 0, count: n * 3)
		var si = 0
		var di = 0
		for _ in 0..<n {
			let r = Int(image.data[si])
			let g = Int(image.data[si + 1])
			let b = Int(image.data[si + 2])
			let a = Int(image.data[si + 3])
			out[di] = UInt8((r * a) / 255)
			out[di + 1] = UInt8((g * a) / 255)
			out[di + 2] = UInt8((b * a) / 255)
			si += 4
			di += 3
		}
		return RGB3(width: image.width, height: image.height, data: out)
	}
}

File: hash/js/rosetta-image-hash/src/internal/imgRgb.ts
/**
 * Normalize an RgbImage (3 or 4 channels) to a flat Uint8Array of RGB triples
 * (row-major, length = width * height * 3).
 *
 * For channels=4, alpha is composited against opaque black via
 * `out_c = floor(src_c * alpha / 255)`, matching PIL `convert('RGB')`.
 * Fully transparent → (0,0,0); fully opaque → src RGB unchanged.
 */

import type { RgbImage } from "../hash.js";

export interface Rgb3 {
  width: number;
  height: number;
  /** Length = width * height * 3, row-major. */
  data: Uint8Array;
}

export function toRgb(img: RgbImage): Rgb3 {
  const { width, height, channels, data } = img;
  if (channels === 3) {
    return { width, height, data };
  }
  // channels === 4
  const out = new Uint8Array(width * height * 3);
  let si = 0;
  let di = 0;
  for (let i = 0; i < width * height; i++) {
    const r = data[si];
    const g = data[si + 1];
    const b = data[si + 2];
    const a = data[si + 3];
    out[di] = Math.trunc((r * a) / 255);
    out[di + 1] = Math.trunc((g * a) / 255);
    out[di + 2] = Math.trunc((b * a) / 255);
    si += 4;
    di += 3;
  }
  return { width, height, data: out };
}
